Changes to /admin page to promote a user to language/question coordinator or to deassign a user from coordinator rights.GSoC M2.3

kshitij01042002 opened the pull request:
## Overview
<!--
READ ME FIRST:
Please answer *both* questions below and check off every point from the Essential Checklist!
If there is no corresponding issue number, fill in N/A where it says [fill_in_number_here] below in 1.
-->

1. This PR fixes #18904 and part of #17440 .
2. This PR does the following: Changes to /admin page to promote a user to language/question coordinator or to deassign a user from coordinator rights. Backend handler changes required for the same.
End-to-End tests for the flow promoting a user to translation coordinator.

## Essential Checklist

- [x] The **PR title** starts with "Fix #bugnum: " or "Fix part of #bugnum: ...",
followed by a short, clear summary of the changes.
- [x] I have followed the [instructions for making a code change](https://github.com/oppia/oppia/wiki/Make-a-pull-request).
- [x] The linter/Karma presubmit checks have passed on my local machine.
- [x] "Allow edits from maintainers" is checked. (See [here](https://help.github.com/en/github/collaborating-with-issues-and-pull-requests/allowing-changes-to-a-pull-request-branch-created-from-a-fork) for instructions on how to enable it.)
- [x] My PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] I have assigned the correct reviewers to this PR (or will leave a
comment with the phrase "@{{reviewer_username}} PTAL" if I don't have
permissions to assign reviewers directly).


## Proof that changes are correct
[Screencast from 24-09-23 04:00:43 PM IST.webm](https://github.com/oppia/oppia/assets/91470808/8c273eef-37cd-4e8b-af1c-ae9774184043)

<!--
Add videos/screenshots of the user-facing interface to demonstrate that the changes
made in this PR work correctly. If this PR is for a developer-facing feature,
provide the videos/screenshots of developer-facing interface. Please also include
videos/screenshots of the developer tools browser console, so that we can be
sure that there are no console errors.

When you make updates to the PR, please update these videos/screenshots as well.
You can drop videos/screenshots from previous versions of the PR.

The above should be done for all PRs, including short ones (e.g. a single-line change).
However, if the changes in your PRs are autogenerated via a script and you cannot
provide proof for the changes then please leave a comment "No proof of changes
needed because {{Reason}}" and remove all the sections below.
-->

#### Proof of changes on desktop with slow/throttled network

<!--
Make sure to properly verify that everything works correctly, and that there are
no weird UI mistakes or other problems. Also, if there are any newly added fields,
try to fill them out and test that different inputs are correctly accepted/rejected.

Throttle the network (to 3G) using the browser Developer Tools (see references below).
There should be no performance or UI issues while the network is slow.

References:
 - Chrome: https://css-tricks.com/throttling-the-network/
 - Firefox: https://developer.mozilla.org/en-US/docs/Tools/Network_Monitor/Throttling
-->

#### Proof of changes on mobile phone

<!--
In some cases this is not needed (e.g. for pages that we do not expect to
support mobile phones, or for backend-only features).

Feel free to use the Developer Tools emulator for this.

References:
 - Chrome: https://developer.chrome.com/docs/devtools/device-mode/
 - Firefox: https://firefox-source-docs.mozilla.org/devtools-user/index.html#responsive-design-mode
-->

#### Proof of changes in Arabic language

<!--
If the PR changes the UI, make sure to add screenshots with the site
language set to Arabic as well (we use Arabic as it is a language written from right to left).
-->

## PR Pointers

- Never force push! If you do, your PR will be closed.
- To reply to reviewers, follow these instructions: https://github.com/oppia/oppia/wiki/Make-a-pull-request#step-5-address-review-comments-until-all-reviewers-approve
- Some e2e tests are flaky, and can fail for reasons unrelated to your PR. We are working on fixing this, but in the meantime, if you need to restart the tests, please check the ["If your build fails" wiki page](https://github.com/oppia/oppia/wiki/If-CI-checks-fail-on-your-PR).
- See the [Code Owner's wiki page](https://github.com/oppia/oppia/wiki/Oppia's-code-owners-and-checks-to-be-carried-out-by-developers) for what code owners will expect.

kshitij01042002 committed:
failing tests

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
restored changed file

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into develop

kshitij01042002 committed:
implemented translation coordinator role and handler changes

kshitij01042002 committed:
fix lint errors

kshitij01042002 committed:
fix mppy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
Merge branch 'develop' of https://github.com/oppia/oppia into gsoc_M2.1

kshitij01042002 committed:
first pass by Sean

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
reviewer changes

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
Sean's comments addressed

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.1

kshitij01042002 committed:
fixed docstring

kshitij01042002 committed:
Merge branch 'gsoc_M2.1' of https://github.com/kshitij01042002/oppia into gsoc_M2.1

kshitij01042002 committed:
first draft pass

kshitij01042002 committed:
second draft

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
fixed e2e ci error

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.2

kshitij01042002 committed:
reviewers commit

kshitij01042002 committed:
Merge branch 'gsoc_M2.2' of https://github.com/kshitij01042002/oppia into gsoc_M2.2

kshitij01042002 committed:
first pass

kshitij01042002 committed:
e2e tests

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed test coverage

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed mypy errors

kshitij01042002 committed:
fixed ts errors

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
covered frontend coverage

kshitij01042002 committed:
Merge remote-tracking branch 'upstream/develop' into gsoc_M2.3

kshitij01042002 committed:
fixed merge conflicts

kshitij01042002 committed:
fixed e2e error

kshitij01042002 committed:
covered backend test coverage

kshitij01042002 commented:
Hi @sagangwee @vojtechjelinek @DubeySandeep @lkbhitesh07 @ShivamJhaa can you PTAL? Thanks.

kshitij01042002 committed:
Update gae_models_test.py

kshitij01042002 committed:
fixed failing testsz

kshitij01042002 committed:
Merge branch 'gsoc_M2.3' of https://github.com/kshitij01042002/oppia into gsoc_M2.3

kshitij01042002 committed:
fixed test coverafe

DubeySandeep commented on core/controllers/admin.py:
```diff
@@ -1309,13 +1310,20 @@ def get(self) -> None:
+            if feconf.ROLE_ID_TRANSLATION_COORDINATOR in user_roles:
+                coordinated_language_ids = [
+                    rights.language_id for rights in
+                    suggestion_services.get_translation_rights_with_user(
```
`get_translation_rights_with_user` function inside `suggestion_services` doesn't sound right to me as the suggestion should not be serving user's permissions/rights. Move it to  rights_services.

DubeySandeep replied:
I took a quick pass an seems like all the permission models/object/services are inside the suggestion context, can we move it to user or rights context?

kshitij01042002 replied:
Since the rights were related to suggestions so I added it in the `suggestion_services`. Moved it to `user_services`. PTAL.

DubeySandeep commented on core/storage/suggestion/gae_models.py:
```diff
@@ -3506,6 +3506,26 @@ class TranslationCoordinatorsModel(base_models.BaseModel):
         indexed=True, required=True)
 
+    # Here we use MyPy ignore because the signature of this method
+    # doesn't match with BaseModel.get().
```
Was there any specific reason for creating a separate model and not using the existing `UserContributionRights` model?

kshitij01042002 replied:
`UserContributionRights` have one instance per user and the new model has one instance per language. We are displaying stats of all the languages at once in the new dashboard so it would have been to many calls to the datastore. [Mock Link](https://www.figma.com/file/bmyOkQwosePcoNAr45Rdk0/Admin-Dashboard?node-id=35%3A3752&mode=dev) to get a better idea of what we are planning to do.

`UserContributionRights` stores contribution rights of the user and the new model stores administrative rights of the user

DubeySandeep reviewed (commented):
@kshitij01042002 I took a quick pass and left two comments/questions, PTAL!

DubeySandeep commented:
Also, are we expecting `translation_admin` to assign `translation_coodinator` or the `superAdmin`? I think it would be better to allow `translation_admin` to handle translation coordinators, isn't it? Maybe I'm missing the requirements, would you mind explaining the reason behind the design choices? Thanks! :)

kshitij01042002 commented:
Hi @DubeySandeep Thanks for the review. With the implementation of this feature we are planning to remove the roles of translation admin and question admin and introduce two new roles transation coordinator and question coordinator. So only super Admin can assign a user to the role of translation/question coordinator.

With the new contributor-admin-dashboard getting launched, we are changing the general translation rights and making it language speicifc i.e. While assigning the role of translation coordinator SuperAdmin have to assign the coordinator to language(s).

Please let me know if you have any more concerns or suggestions.

Reference: [GSoC Proposal](https://docs.google.com/document/d/19lCEYQUgV7_DwIK_0rz3zslRHX2qKOHn-t9Twpi0qu0/edit).

kshitij01042002 committed:
shifted services from suggestion to user

kshitij01042002 committed:
removed unused imports

kshitij01042002 committed:
fixed import errors

lkbhitesh07 commented on core/controllers/admin_test.py:
```diff
@@ -1910,6 +1920,272 @@ def test_adding_new_topic_manager_to_a_topic(self) -> None:
+            })
+        self.logout()
+
+    def test_adding_two_translation_coordinators_to_language(self) -> None:
```
This should be `test_assigning_two_languages_to_translation_coordinator`?

kshitij01042002 replied:
Yeah you are correct. Thanks for the nit.

lkbhitesh07 reviewed (changes requested):
Thanks for the changes @kshitij01042002. One minor nit. I have one question, I took a look at the video proof you provided in the PR description, and I noticed that when you initially added `Akan` as the language and then visited dsadmin I could only see the entry for the `en` part and not for the `Akan` part, further when you add more languages then I'm able to see `ak` language code entry which I'm guessing is for `Akan` but there is no entry for the last language. Can you please explain this a bit. Let me know if I'm missing something.

Thanks!

kshitij01042002 commented:
> Thanks for the changes @kshitij01042002. One minor nit. I have one question, I took a look at the video proof you provided in the PR description, and I noticed that when you initially added `Akan` as the language and then visited dsadmin I could only see the entry for the `en` part and not for the `Akan` part, further when you add more languages then I'm able to see `ak` language code entry which I'm guessing is for `Akan` but there is no entry for the last language. Can you please explain this a bit. Let me know if I'm missing something.
> 
> Thanks!

Hi @lkbhitesh07 I think you miss understood the dropdown option as the selected option. The language is only added when the user clicks on the `+` button. **Akan** that you mention is the option available to add next. Whatever languages you see under title `User has rights in following languages` are the once that are assigned to the user. Please let me know if there's still any confusion.

kshitij01042002 committed:
minor nit

kshitij01042002 commented:
Hi @DubeySandeep and @lkbhitesh07 I have answered your review comments. PTAL. Thanks for the review.

kshitij01042002 committed:
moved coordinator domain object from suggestion to user domain

kshitij01042002 committed:
fixed type errors

lkbhitesh07 reviewed (approved):
Thanks for the explanation. LGTM for the codeowner files.

sagangwee commented:
@StephenYu2018 Hey Stephen, can you take a first pass here? Thanks.

sagangwee reviewed (approved):
@kshitij01042002 LGTM for codeowners.

StephenYu2018 commented:
@sagangwee Sorry, I was away yesterday. Can I still take a pass on this PR?

seanlip commented:
@StephenYu2018 Go for it -- in general you're welcome to take a pass on any PR!

StephenYu2018 commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6169,3 +6169,39 @@ def test_get_field_name_mapping_to_takeout_keys(self) -> None:
+            )
+        )
+
+        self.assertEqual(
+            len(translation_coordinators_models),
+            2
+        )
```
This test case reads "When getting the translation coordinators models by a given user, then the result should have two models", which feels incomplete. Considering the behavior we're trying to test, I feel that the test should also explicitly check that all `TranslationCoordinatorsModel` instances that came from the query should have the same user ID. Then, it'll read something like "When getting the translation coordinators models by a given user, then the result should have two models, both with the same user", which feels much more complete.

kshitij01042002 replied:
Sounds about right, I have added it PTAL.

StephenYu2018 replied:
Not quite. When we want all models that have a specific user, then after we've made the query to storage, we expect all the returned models to have the same user we've specified. So with the assert statements you've added:

```python
        self.assertEqual(
            translation_coordinators_models[0].id,
            self.LANGUAGE_4_ID
        )
        self.assertEqual(
            translation_coordinators_models[1].id,
            self.LANGUAGE_5_ID
        )
```

We don't expect nor care that the first model is for language `LANGUAGE_4_ID`, or that the second model is for language `LANGUAGE_5_ID`. These assert statements should be removed and replaced with the following:

```python
        self.assertIn(
            self.USER_ID_2
            translation_coordinators_models[0].coordinator_ids,
        )
        self.assertIn(
            self.USER_ID_2
            translation_coordinators_models[1].coordinator_ids,
        )
```

kshitij01042002 replied:
Ohh Yes, you are correct. Changed it. PTAL again.

StephenYu2018 commented on core/domain/user_services_test.py:
```diff
@@ -4144,3 +4147,200 @@ def test_remove_question_submit_rights(self) -> None:
+        self.set_translation_coordinators(['C'], 'hi')
+        self.set_translation_coordinators(['C'], 'fr')
+        self.assertEqual(3, len(
+            user_services.get_translation_rights_with_user(user_id_c)))
```
If we're trying to de-assign a user from all languages, does it matter that we start out with *specifically* 3 languages? I think that any discrete number of languages would be fine. It isn't important to explicitly check that our setup for this test resulted in the user being assigned 3 languages. At best, it's pointless and at worst, it confuses the reader about what is important to this test's results.

kshitij01042002 replied:
If not 3, we atleast need to assign 2 languages to the user cause this test is different from the one where we deassign user from one language at a time. I have reduced it to 2. How does that sounds to you?

StephenYu2018 replied:
Sorry this thread was hidden to me. My response is in my second review.

StephenYu2018 commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6169,3 +6169,39 @@ def test_get_field_name_mapping_to_takeout_keys(self) -> None:
+            translation_coordinators_model.id,
+            self.LANGUAGE_4_ID
+        )
+        self.assertEqual(
+            translation_coordinators_model.coordinator_ids,
+            [self.USER_ID_2]
+        )
+        self.assertEqual(
+            translation_coordinators_model.coordinators_count,
+            1
+        )
```
I think these 2 assert statements don't need to be there. It's enough to check that we got the same exact model by matching the model IDs. By removing the 2 assert statements, the test will make it abundantly clear to the reader that each `TranslationCoordinatorModel` is identified by its language.

kshitij01042002 replied:
Sg. Thanks.

StephenYu2018 commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.spec.ts:
```diff
@@ -382,6 +410,62 @@ describe('Admin roles tab component ', function() {
+      ngbModal = TestBed.inject(NgbModal);
+    });
+
+    it('should open the TranslationCoordinatorrRoleEditor', fakeAsync(() => {
```
Typo: Extra `r` after Coordinator

kshitij01042002 replied:
Thanks for the nit. Fixed.

StephenYu2018 replied:
Sorry I should've clarified that all instances of the method name with this typo should be fixed.

StephenYu2018 reviewed (commented):
Took a pass.

kshitij01042002 committed:
review comments by Stephen

kshitij01042002 commented:
Hi @StephenYu2018 I have addressed all your comments. PTAL. Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

qinghaoyang reviewed (approved):
@kshitij01042002 LGTM

vojtechjelinek commented on core/domain/user_domain.py:
```diff
@@ -1651,3 +1651,37 @@ def validate(self) -> None:
+
+
+class TranslationCoordinatorStatsDict(TypedDict):
+    """Dict representation of TranslationCoordinatorStats domain object"""
```
```suggestion
    """Dict representation of TranslationCoordinatorStats domain object."""
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_domain_test.py:
```diff
@@ -1694,3 +1694,24 @@ def test_validation(self) -> None:
             'they are already its learner.')
+
+
+class TranslationCoordinatorStatsUnitTests(
+    test_utils.GenericTestBase):
```
```suggestion
class TranslationCoordinatorStatsUnitTests(
    test_utils.GenericTestBase
):
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_domain_test.py:
```diff
@@ -1694,3 +1694,24 @@ def test_validation(self) -> None:
+    }
+
+    def test_to_dict(self) -> None:
+        actual_stats = user_domain.TranslationCoordinatorStats( # pylint: disable=line-too-long
```
Do not sue pylint disable.

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -2995,3 +2995,223 @@ def is_user_blog_post_author(user_id: str) -> bool:
+            'Guest user is not allowed to assign roles to a user.'
+        )
+
+    if (role_services.ACTION_MODIFY_CORE_ROLES_FOR_ANY_ACTIVITY not in
+            committer.actions):
```
```suggestion
    if (
        role_services.ACTION_MODIFY_CORE_ROLES_FOR_ANY_ACTIVITY not in
        committer.actions
    ):
```

vojtechjelinek replied:
ditto below

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -2995,3 +2995,223 @@ def is_user_blog_post_author(user_id: str) -> bool:
+        list(TranslationCoordinatorStats). The rights objects associated with
+        the languagesassigned to given user.
+    """
+    translation_coordinator_models: Sequence[
+        suggestion_models.TranslationCoordinatorsModel] = (
+        suggestion_models.TranslationCoordinatorsModel.get_by_user(user_id))
+    return [
+        get_translation_rights_from_model(model)
+        for model in translation_coordinator_models
+        if model is not None]
```
```suggestion
    translation_coordinator_models: Sequence[
        suggestion_models.TranslationCoordinatorsModel
    ] = suggestion_models.TranslationCoordinatorsModel.get_by_user(user_id)
    
    return [
        get_translation_rights_from_model(model)
        for model in translation_coordinator_models
        if model is not None
    ]
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services_test.py:
```diff
@@ -4144,3 +4147,199 @@ def test_remove_question_submit_rights(self) -> None:
+            self.get_user_id_from_email(self.CURRICULUM_ADMIN_EMAIL))
+
+        self.set_curriculum_admins([self.CURRICULUM_ADMIN_USERNAME])
+        self.set_translation_coordinators(
+            [user_services.get_username(
+                self.user_id_translationcoordinator)], 'en')
```
```suggestion
        self.set_translation_coordinators([
            user_services.get_username(
                self.user_id_translationcoordinator
            )
        ], 'en')
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services_test.py:
```diff
@@ -4144,3 +4147,199 @@ def test_remove_question_submit_rights(self) -> None:
+
+        with self.assertRaisesRegex(
+            Exception,
+            'UnauthorizedUserException: Could not assign new role.'):
```
```suggestion
            'UnauthorizedUserException: Could not assign new role.'
        ):
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services_test.py:
```diff
@@ -4144,3 +4147,199 @@ def test_remove_question_submit_rights(self) -> None:
+                self.user_admin, guest_user, 'en')
+
+    def test_reassigning_role_to_same_user(self) -> None:
+        with self.assertRaisesRegex(
+            Exception, 'This user already is a coordinator for this language'):
```
```suggestion
        with self.assertRaisesRegex(
            Exception, 'This user already is a coordinator for this language'
        ):
```

vojtechjelinek replied:
Fix this also below

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/storage/suggestion/gae_models.py:
```diff
@@ -3576,3 +3596,18 @@ def export_data(cls, user_id: str) -> Dict[str, List[str]]:
         }
+
+    @classmethod
+    def get_by_user(cls, user_id: str) -> Sequence[
+        TranslationCoordinatorsModel]:
```
```suggestion
    def get_by_user(cls, user_id: str) -> Sequence[
        TranslationCoordinatorsModel
    ]:
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/storage/suggestion/gae_models_test.py:
```diff
@@ -6169,3 +6169,39 @@ def test_get_field_name_mapping_to_takeout_keys(self) -> None:
                 'coordinator_ids': 'coordinated_language_ids'
             })
+
+    def test_get_returns_model_when_it_exists(self) -> None:
+
```
```suggestion
    def test_get_returns_model_when_it_exists(self) -> None:
```

vojtechjelinek replied:
ditto below

kshitij01042002 replied:
Done.

vojtechjelinek reviewed (changes requested):
Thanks! Took a pass.

DubeySandeep commented on core/tests/webdriverio_utils/AdminPage.js:
```diff
@@ -213,6 +213,41 @@ var AdminPage = function() {
+    await action.matSelect(
+      'New role selector', roleSelector, 'translation coordinator');
+
+    var languageSelectorModal = $('.e2e-test-language-selector-modal');
+    var languageSelector = $('.e2e-test-language-selector');
+    var languageSelectorCloseButton = $(
+      '.e2e-test-language-selector-close-button');
+    var languageSelectorAddButton = $('.e2e-test-language-selector-add-button');
```
Define these on the top*

kshitij01042002 replied:
Done. Thanks for the review.

DubeySandeep reviewed (approved):
LGTM for the codeowner file, left one minor comment, PTAL!

kshitij01042002 committed:
review comments by vojtech

kshitij01042002 committed:
review comments by sandeeo

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

kshitij01042002 commented:
Hi @vojtechjelinek I have addressed all your comments. can you PTAL again?

Thanks for the review.

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -2995,3 +2995,232 @@ def is_user_blog_post_author(user_id: str) -> bool:
+    )
+
+
+def get_translation_rights_with_user(user_id: str) -> List[
+    user_domain.TranslationCoordinatorStats]:
```
```suggestion
def get_translation_rights_with_user(user_id: str) -> List[
    user_domain.TranslationCoordinatorStats
]:
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/domain/user_services.py:
```diff
@@ -2995,3 +2995,232 @@ def is_user_blog_post_author(user_id: str) -> bool:
+    Returns:
+        bool. True if the user is coordinator or else False.
+    """
+    model = suggestion_models.TranslationCoordinatorsModel.get(
+            language_id)
```
```suggestion
    model = suggestion_models.TranslationCoordinatorsModel.get(
        language_id)
```

kshitij01042002 replied:
Done.

vojtechjelinek commented on core/storage/suggestion/gae_models.py:
```diff
@@ -3506,6 +3506,26 @@ class TranslationCoordinatorsModel(base_models.BaseModel):
+    # Here we use MyPy ignore because the signature of this method
+    # doesn't match with BaseModel.get().
+    # https://mypy.readthedocs.io/en/stable/error_code_list.html#check-validity-of-overrides-override
+    @classmethod
+    def get( # type: ignore[override]
+        cls, language_code: str
+    ) -> Optional[TranslationCoordinatorsModel]:
+        """Gets the TranslationCoordinatorsModel
+        matching the supplied language_code.
+
+        Args:
+            language_code: str. ISO 639-1 language code.
+
+        Returns:
+            TranslationCoordinatorsModel|None. The matching
+            TranslationCoordinatorsModel, or None if no
+            such model instance exists.
+        """
+        return cls.get_by_id(language_code)
+
```
Why do you even need to implement this? `get` is implemented in the base model

kshitij01042002 replied:
We need to override the default method in base model cause if there's no model found for given id default method fails with an error but this method returns None instead. We need this functionality in the case where we check if the model exists and if it doesn't exists we create a new one. Please let me know if there's any concerns.

vojtechjelinek replied:
You can do `.get(id, strict=False)` and that will return either the model or `None`, you do not need this reimplementation.

kshitij01042002 replied:
Removed it. Thanks for the information. PTAL again.

vojtechjelinek reviewed (changes requested):
Thanks! Few other comments.

kshitij01042002 committed:
pass 2 by vojtech

kshitij01042002 committed:
Merge branch 'gsoc_M2.3' of https://github.com/kshitij01042002/oppia into gsoc_M2.3

kshitij01042002 commented:
Hi @vojtechjelinek I have addressed all your comments please take a look. Thanks.

kshitij01042002 committed:
added back the get method

kshitij01042002 committed:
added period nit

StephenYu2018 commented on core/tests/test_utils.py:
```diff
@@ -2555,6 +2555,23 @@ def set_topic_managers(
                     'topic_id': topic_id
                 }, csrf_token=self.get_new_csrf_token())
 
+    def set_translation_coordinators(
+        self, translation_coordinator_usernames: List[str], language_id: str
+    ) -> None:
+        """Sets role of given users as TRANSLATION_COORDINATOR.
+
+        Args:
+            translation_coordinator_usernames: list(str). List of usernames.
+            language_id: str. The language Id.
+        """
+        with self.super_admin_context():
+            for username in translation_coordinator_usernames:
+                self.put_json('/translationcoordinatorrolehandler', {
+                    'username': username,
+                    'action': 'assign',
+                    'language_id': language_id
+                }, csrf_token=self.get_new_csrf_token())
```
This method needs to be covered by tests. From what I've read, the features we want to provide by adding this method includes:
- Our testing utilities `gives_translation_coordinator_role_in_a_language_to_users`
  - **Motivation of scenario**: Users that are not translation coordinators will become translation coordinators 
  - **Action**: Giving translation coordinator role in a language to users without the role
  - **Expected result**: All mentioned users are translation coordinators in the given language 
- Our testing utilities `adds_another_translation_coordinator_language_to_translation_coordinator`
  - **Motivation of scenario**: Translation coordinator(s) can be assigned new languages without replacing their existing ones
  - **Action**: Giving a unique translation coordinator language multiple times
  - **Expected result**: The user(s) should have all of the given translation coordinator languages

So having 2 test cases that cover each of the aforementioned features should be adequate.

kshitij01042002 replied:
Yeah, you are correct. Sorry I missed out on writing test for this code. Thanks for looking out.

StephenYu2018 commented on core/domain/user_services_test.py:
```diff
@@ -4144,3 +4147,204 @@ def test_remove_question_submit_rights(self) -> None:
+        user_id_c = self.get_user_id_from_email('c@example.com')
+        self.set_translation_coordinators(['C'], 'en')
+        self.set_translation_coordinators(['C'], 'hi')
+        self.assertEqual(2, len(
+            user_services.get_translation_rights_with_user(user_id_c)))
```
Remove this assert statement. If we want to ensure that calling `set_translation_coordinators` twice on the same user(s) assigns said user(s) with both translation coordinator languages, then write a separate test for `set_translation_coordinators`. For details about this test, please see my review comment in the `core/tests/test_utils.py` file.

kshitij01042002 replied:
Sounds about right. Done.

StephenYu2018 commented on core/templates/pages/admin-page/roles-tab/admin-roles-tab.component.ts:
```diff
@@ -124,12 +131,38 @@ export class AdminRolesTabComponent implements OnInit {
     });
   }
 
+  openTranslationCoordinatorrRoleEditor(): void {
```
Again, please fix the typo. It should be `openTranslationCoordinatorRoleEditor`.

kshitij01042002 replied:
Done. Thanks for the nit. Sorry for not checking for other typos as well.

StephenYu2018 reviewed (commented):
I've left 4 comments.

kshitij01042002 committed:
comments by Stephen and vojtech

kshitij01042002 commented:
Hi @StephenYu2018 and @vojtechjelinek I have addressed all your comments. PTAL again thanks.


Re @StephenYu2018 If possible can you resolve the comments which seems satisfactory to you? It helps me track all the open comments. Thanks.

vojtechjelinek reviewed (approved):
LGTM!

StephenYu2018 reviewed (approved):
LGTM! Thanks.

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
fixed lint errors

kshitij01042002 committed:
Merge branch 'develop' into gsoc_M2.3
